Controller updates for Panda Linking

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR implements the initial updates to the controller to handle panda linking.  This change adds:

* `panda_link_table` which holds all instances of panda links and their configurations
* `panda_link_frontend` and `panda_link_backend` that handle upserting and removing panda links from the table
* Added controller commands to upsert and remove panda links
* Added controller RPC messages for upserting and removing panda links

Fixes: CORE-10222

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
cluster/panda_link: Introduce types

Added initial configuration settings for Panda Links

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/panda_link: Introduce table

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
model: Added panda_link batch type

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Added panda_link commands

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/snapshot: Added panda link to snapshot

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Added rpc commands for panda_link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/rpc: Added panda link rpc handlers

Placeholders that return success until full implementation in place.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Added panda link error codes

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Added panda link frontend

Added frontend service that handles commands to add or remove panda
links from the panda link table.

Added tests to validate the validator

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Wire in panda link into the controller

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--8c5ffe44-090f-4efd-a99b-f4d558e7f0b9-->
<!--TEST RESULTS-->
<!--8bf771b6-e25d-4275-a40a-9e0ee76fc472-->
<!--TEST RESULTS-->
<!--2606de0b-8a9a-40d5-837c-760aa77c3e62-->
<!--TEST RESULTS-->
<!--0bfc817e-e652-42e8-ac6b-479f8108c4d4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66649">build#66649</a></summary>


| test_class               | test_method               | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:--------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| raft_fixture             | leadership_transfer_delay |                                                                                                                                     | unit        | https://buildkite.com/redpanda/redpanda/builds/66649#0197244f-fc48-4e41-a1d1-12187a774c40 | FAIL          | 0/1      |                                                                                                                                                                        |
| CloudRetentionTest       | test_cloud_retention      | {"cloud_storage_type": 2, "max_consume_rate_mb": null}                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66649#019724c3-a82d-43a0-a84b-ed3ef289fc1c | FLAKY         | 20/21    | upstream reliability is '97.5'. current run reliability is '95.23809523809523'. drift is 2.2619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest | test_node_operations      | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66649#019724c3-a82e-4665-8f5a-e96fb9b679e8 | FLAKY         | 19/21    | upstream reliability is '98.30508474576271'. current run reliability is '90.47619047619048'. drift is 7.82889 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66898">build#66898</a></summary>


| test_class               | test_method                                          | test_arguments                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:-----------------------------------------------------|:-------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite          | test_cpu_profiler_enable_override_filter_old_samples |                                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/66898#01973dd4-2ed9-45f5-b831-2d10b3516ac6 | FAIL          | 0/1      |                                                                                                                                                                         |
| raft_fixture             | leadership_transfer_delay                            |                                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/66898#01973d7c-632e-414f-9b84-29e45c7cc647 | FAIL          | 0/1      |                                                                                                                                                                         |
| NodePostRestartProbeTest | post_restart_probe_test                              |                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66898#01973e96-a3aa-4856-a85c-92c3f11d198c | FLAKY         | 20/21    | upstream reliability is '96.73913043478261'. current run reliability is '95.23809523809523'. drift is 1.50104 and the allowed drift is set to 50. The test should PASS  |
| TestTieredStoragePause   | test_safe_pause_resume                               | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66898#01973e96-a3ab-400f-abca-17a4883b52d0 | FLAKY         | 20/21    | upstream reliability is '89.69072164948454'. current run reliability is '95.23809523809523'. drift is -5.54737 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66989">build#66989</a></summary>


| test_class                  | test_method                              | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66989#019745ed-3329-4c86-aed0-c36aa4ea8615 | FLAKY         | 19/21    | upstream reliability is '97.0059880239521'. current run reliability is '90.47619047619048'. drift is 6.5298 and the allowed drift is set to 50. The test should PASS   |
| MaintenanceTest             | test_maintenance_sticky                  | {"use_rpk": false}                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66989#019745ed-3327-4e90-9530-15f8d926f2f1 | FLAKY         | 20/21    | upstream reliability is '96.88581314878893'. current run reliability is '95.23809523809523'. drift is 1.64772 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66989#019745ed-3329-4c86-aed0-c36aa4ea8615 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest    | test_node_operations                     | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66989#019745ed-332a-4a4b-9009-5c2a536bac90 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67056">build#67056</a></summary>


| test_class               | test_method                               | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePoolMigrationTest    | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "off", "cleanup_policy": "compact", "test_mode": "tiered_storage"}                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67056#019755a9-ee8a-4308-b11d-a4199a898491 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations                      | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67056#019755a9-ee88-4760-b7b9-91b4b36dc960 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations                      | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67056#019755a9-ee87-4481-b696-aa88f13955b1 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS            |
| DisablingPartitionsTest  | test_disable                              |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67056#0197558b-00e3-4516-b8d4-39eed0d11720 | FLAKY         | 18/21    | upstream reliability is '91.47121535181236'. current run reliability is '85.71428571428571'. drift is 5.75693 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented on src/v/model/BUILD:
```diff
@@ -36,6 +37,7 @@ redpanda_cc_library(
         "metadata.h",
         "namespace.h",
         "offset_interval.h",
+        "panda_link.h",
```
does this have to be in model ? I think it may be enough to put it to `cluster/panda_link` or create a completely separate submodule

michael-redpanda replied:
When do types end up in `model` and when don't they?  Can you point to design documentation that provides guidance?

michael-redpanda replied:
Chatted offline with Michal and we are aligned on path forward here

rockwotj replied:
We need to stop the mega model target/package. It does not scale the build well. I can outline some approaches I have seen scale better, but honestly I think everyone struggles with this

rockwotj replied:
+1 to new link submodule

mmaslankaprv commented on src/v/model/panda_link.h:
```diff
@@ -0,0 +1,134 @@
+
+#include <seastar/util/bool_class.hh>
+
+namespace model {
```
if we put it into separate namespace then we can skip repeating `panda_link...`

mmaslankaprv commented on src/v/model/panda_link.h:
```diff
@@ -0,0 +1,134 @@
+/**
+ * @brief Represents the settings for connection to a remote cluster
+ */
+struct panda_link_connection
```
here the name suggest that the object represents connection, while it is only representing configuration, consider changing the name to `connection_configuration`

michael-redpanda replied:
Fair point, i'll change it

mmaslankaprv commented on src/v/model/panda_link.h:
```diff
@@ -0,0 +1,134 @@
+    /// Authentication configuration for the connection
+    std::optional<authn_variant> authn_config;
+    /// Path to certificate file to use
+    ss::sstring cert_file_path;
+    /// Path to key file (when mTLS is in use)
+    ss::sstring key_file_path;
+    /// Path to the CA file to use
+    ss::sstring ca_file_path;
```
Consider representing those as bytes containing the certificate data, paths are not k8s friendly.

michael-redpanda replied:
should we support both file and buffer based?

michael-redpanda replied:
this will require changes for our client to support buffer-based keys/certs - definitely doable but at the moment it only supports files.  I'll file a follow up work item to add in support for buffer based keys/certs

rockwotj replied:
+1 to doing a non-file based approach, honestly I would only do that. Having a config that points to the same file on every disk is a bit funky. But I get it works for ansible

michael-redpanda replied:
Definitely agreed - filed [CORE-11926](https://redpandadata.atlassian.net/browse/CORE-11926) to update the k/client to accept buffers

mmaslankaprv commented on src/v/cluster/panda_link_table.h:
```diff
@@ -0,0 +1,57 @@
+ */
+class panda_link_table {
+public:
+    using map_t
+      = absl::btree_map<model::panda_link_id, model::panda_link_metadata>;
```
do we need this data to be sorted ?

michael-redpanda replied:
per discussion will convert this to a chunked_map of some sort

mmaslankaprv commented on src/v/cluster/panda_link_table.h:
```diff
@@ -0,0 +1,57 @@
+        bool operator()(const model::panda_link_name&, std::string_view) const;
+    };
+
+    using name_index_t = absl::
+      btree_map<model::panda_link_name, model::panda_link_id, name_less_cmp>;
```
are we relying on the fact that index is sorted ?

mmaslankaprv commented on src/v/cluster/panda_link_backend.cc:
```diff
@@ -0,0 +1,58 @@
+  : _table(table) {}
+
+ss::future<std::error_code>
+panda_link_backend::apply_update(model::record_batch b) {
```
Usually all other `cluster` elements use `backend` as a component that executes the asynchronous actions that are driven by the updates of the table. Consider implementing `apply_update` as a method of the `table` class.

michael-redpanda replied:
The `transform` system, which I based a bit of this off of, uses this pattern.  What is the issue with this pattern?

rockwotj replied:
The transform subsystem never exposes the table outside the backend and frontend, so it's okay IMO and makes testing easier

rockwotj replied:
That being said - it not clear what the pattern here will be.

mmaslankaprv commented on src/v/cluster/panda_link_table.h:
```diff
@@ -36,6 +36,25 @@ class panda_link_table {
+    std::optional<model::panda_link_id>
+    find_id_by_name(const model::panda_link_name& name) const;
+
+    /// Upserts a link, if the ID classes, throws a std::logic_error
+    void upsert_link(model::panda_link_id, model::panda_link_metadata);
+    /// Removes a link by ID
+    void remove_link(const model::panda_link_name&);
```
Those two methods should be private, the only way link table can be updated should be by applying the command

mmaslankaprv commented on src/v/cluster/panda_link_backend.h:
```diff
@@ -0,0 +1,33 @@
+public:
+    explicit panda_link_backend(ss::sharded<panda_link_table>*);
+
+    ss::future<std::error_code> apply_update(model::record_batch);
+    bool is_batch_applicable(const model::record_batch&) const;
+
+    ss::future<> fill_snapshot(controller_snapshot&) const;
+    ss::future<> apply_snapshot(model::offset, const controller_snapshot&);
```
Consider moving those methods to `table`

michael-redpanda committed:
feature: Added panda_linking_dr to feature table

Signed-off-by: Michael Boquard <michael@redpanda.com>

rockwotj commented on src/v/model/panda_link.h:
```diff
@@ -0,0 +1,134 @@
+      serde::version<0>,
+      serde::compat_version<0>> {
+    /// Name of the panda link
+    panda_link_name name;
```
why does a link need a name? are we supporting multiple?

michael-redpanda replied:
Not initially but I'd imagine we will so having a nice human readable name seems like a good idea

rockwotj replied:
If we have multiple I'm OK with a name - but I also think it could be added later too (ie the initial name would be `default` or something. I also think that means we could remove the name index and some of the generality/complexity from the table since there is just a single one? If you are going to use the notification pattern like transforms it will save some logic in reconciliation to just assume one for now in the code.

To be honest I don't see any cases, even with storage where we can't go from 1 -> N even if the snapshot is a single we can just convert it to a single map entry later on. Supporting multiple feels a bit like YAGNI here

rockwotj commented on src/v/model/panda_link.h:
```diff
@@ -0,0 +1,134 @@
+    /// Connection settings for the panda link
+    panda_link_connection connection;
+    /// Type to indicate if the panda link is paused
+    using panda_link_paused_t = ss::bool_class<struct panda_link_paused_tag>;
```
I think this can just be `paused_t` because it will be referenced as `panda_link_metadata::paused_t`

rockwotj commented on src/v/cluster/panda_link_table.cc:
```diff
@@ -49,4 +55,82 @@ bool panda_link_table::name_less_cmp::operator()(
+        if (name_it->second != id) {
+            throw std::logic_error(fmt::format(
+              "panda link id={} is attempting to use a name {} which is "
+              "already registerd to {}",
```
nit: typo

rockwotj commented on src/v/cluster/panda_link_table.cc:
```diff
@@ -49,4 +55,82 @@ bool panda_link_table::name_less_cmp::operator()(
+    if (it == _underlying.end()) {
+        return std::nullopt;
+    }
+    return std::cref(it->second);
```
absl::btree does not have pointer stability, this can cause dangling pointers. Just make a copy.

rockwotj commented on src/v/cluster/panda_link_frontend.cc:
```diff
@@ -0,0 +1,284 @@
+                max_name_size);
+              return errc::panda_link_invalid_create;
+          }
+          if (!is_valid_utf8(cmd.value.name())) {
+              vlog(
+                clusterlog.info,
+                "Attempting to create a panda link with an invalid UTF-8 name");
+              return errc::panda_link_invalid_create;
+          }
+          if (contains_control_character(cmd.value.name())) {
+              vlog(
+                clusterlog.info,
+                "Attempting to create a panda link with a name containing "
+                "control characters");
+              return errc::panda_link_invalid_create;
+          }
```
let's be more strict and only allow names that are `[A-Za-z0-9_-]+`

rockwotj replied:
I regret not doing that in transforms

michael-redpanda replied:
yeah fair point i'll do that in the next round of updates

michael-redpanda commented:
Force push:

* Moved types out of model and put them in `cluster/panda_link`
* Refactored around model move
* Removed backend - now using table entirely
* Other random fixes

michael-redpanda commented:
I see that C++ lint is failing probably due to the compiler upgrade - I will rebase off of `dev` after the next round of reviews

michael-redpanda commented on src/v/cluster/BUILD:
```diff
@@ -360,10 +360,11 @@ redpanda_cc_library(
         "node/local_monitor.cc",
         "node/types.cc",
         "node_isolation_watcher.cc",
-        "panda_link/types.cc",
```
must have missed a bazel lint earlier

mmaslankaprv commented on src/v/cluster/BUILD:
```diff
@@ -360,6 +360,7 @@ redpanda_cc_library(
         "node/local_monitor.cc",
         "node/types.cc",
         "node_isolation_watcher.cc",
+        "panda_link/types.cc",
```
consider putting it into separate library

rockwotj replied:
Yes please do have a separate target, we should not add to the monolith.

We should probably have the following targets:

`//src/v/cluster/panda_link/models`
`//src/v/cluster/panda_link/controller`

Although I guess the controller bits will be harder to tease out so let's not shave that yak yet, and just make a panda_link models target.

michael-redpanda replied:
So what should the namespace names be?  `cluster::panda_link::controller`?  That seems overly verbose.  Should the namespaces reflect the directory location or not?

mmaslankaprv commented on src/v/cluster/panda_link/table.cc:
```diff
@@ -0,0 +1,158 @@
+              id,
+              meta.name(),
+              name_it->second);
+            return cluster::errc::panda_link_service_error;
```
i thin the error code can be more specific here

mmaslankaprv commented on src/v/cluster/panda_link/table.cc:
```diff
@@ -0,0 +1,158 @@
+    if (name_it != _name_index.end()) {
+        if (name_it->second != id) {
+            vlog(
+              cluster::clusterlog.error,
```
this would be caused by the operator mistake, we can log it with `warn` level, also consider introducing a logger for panda_link subsystem

michael-redpanda replied:
There will be a new logger in `v/panda_link` but are you suggesting a `v/cluster/panda_link/logger` in addition?

michael-redpanda replied:
> this would be caused by the operator mistake

No, that whould be caught much earlier, this represents an error internally

mmaslankaprv commented on src/v/cluster/panda_link/table.cc:
```diff
@@ -0,0 +1,158 @@
+}
+
+cluster::errc table::remove_link(const name_t& name) {
+    auto name_it = _name_index.find(name);
+    if (name_it == _name_index.end()) {
+        return cluster::errc::success;
+    }
```
This is always a question if the delete should be idempotent, i am wondering if it would be better to remove an error if somebody tries to delete a non existing link.

mmaslankaprv commented on src/v/cluster/panda_link/frontend.cc:
```diff
@@ -0,0 +1,297 @@
+          std::move(cmd),
+          timeout - model::timeout_clock::now());
+    }
+    if (ss::this_shard_id() != cluster::controller_stm_shard) {
+        co_return co_await container().invoke_on(
+          cluster::controller_stm_shard,
+          [cmd = std::move(cmd), timeout](auto& service) mutable {
+              return service.do_mutation(std::move(cmd), timeout);
+          });
+    }
+
+    co_return co_await do_local_mutation(std::move(cmd), timeout);
```
looks like this can be simplified to: 
```
co_return co_await container().invoke_on(
    cluster::controller_stm_shard,
    [cmd = std::move(cmd), timeout](auto& service) mutable {
        return service.do_local_mutation(std::move(cmd), timeout);
    });
```

rockwotj replied:
I always find it sneaky that `invoke_on` handles this case, but SGTM

mmaslankaprv commented on src/v/cluster/panda_link/frontend.cc:
```diff
@@ -0,0 +1,297 @@
+    auto b = std::visit(
+      [](auto cmd) { return serde_serialize_cmd(std::move(cmd)); },
+      std::move(cmd));
+    ;
```
nit: redundant semicolon

mmaslankaprv commented on src/v/cluster/panda_link/frontend.cc:
```diff
@@ -0,0 +1,297 @@
+}
+
+cluster::errc frontend::validate_mutation(const panda_link_cmd& cmd) const {
+    static constexpr size_t max_links = 1;
```
this requires a comment

mmaslankaprv commented on src/v/cluster/errc.h:
```diff
@@ -99,6 +99,11 @@ enum class errc : int16_t {
     resource_is_being_migrated,
     invalid_target_node_id,
     topic_id_already_exists,
+    panda_link_does_not_exist,
+    panda_link_invalid_create,
+    panda_link_invalid_update,
+    panda_link_limit_exceeded,
+    panda_link_service_error,
```
maybe we could put these into `panda_link/errc.h` ?

michael-redpanda replied:
So these are specific to errors that occur when modifying the table via the controller so I believe these should stay in the cluster::errc.

There _will_ be a `panda_link::errc` when the actual service is created, but for now these should stay here.

rockwotj commented on src/v/cluster/controller_snapshot.h:
```diff
@@ -14,6 +14,7 @@
 #include "cluster/client_quota_serde.h"
 #include "cluster/cluster_recovery_state.h"
 #include "cluster/data_migration_types.h"
+#include "cluster/panda_link/table.h"
```
nit: consider limiting includes here to just panda_link/types.h and redeclare `map_t` in the snapshot.

rockwotj commented on src/v/cluster/panda_link/table.h:
```diff
@@ -0,0 +1,70 @@
+    bool is_batch_applicable(const model::record_batch&) const;
+    ss::future<std::error_code> apply_update(model::record_batch);
+
+    ss::future<> fill_snapshot(cluster::controller_snapshot&) const;
+    ss::future<>
+    apply_snapshot(model::offset, const cluster::controller_snapshot&);
```
This is now a circular dependency between controller_snapshot and the table. I would prefer if we could make it unidirectional.

rockwotj replied:
A lot of complexity could be remove if the table was just a single optional link for now

dotnwat commented on src/v/cluster/controller_snapshot.h:
```diff
@@ -296,6 +297,16 @@ struct data_migrations_t
+struct panda_link_t
+  : public serde::
+      envelope<panda_link_t, serde::version<0>, serde::compat_version<0>> {
+    panda_link::table::map_t links;
```
since panda_link_t is a serialized data type, is there value in using an explicit map type here rather than a typedef?

michael-redpanda replied:
yeah and I think Tyler correctly pointed out a circular reference here so I'll be changing this up

dotnwat commented on src/v/cluster/panda_link/fwd.h:
```diff
@@ -0,0 +1,16 @@
+
+namespace panda_link {
+class table;
+}
```
should we continue using the `fwd.h` header pattern? i'm generally in favor of just making `panda_link/table.h` be light weight enough to not worry about including it. it's also slightly awkward to decide what bazel rule it should be in..

dotnwat commented on src/v/cluster/panda_link/table.h:
```diff
@@ -0,0 +1,70 @@
+    /// Removes a link by ID
+    cluster::errc remove_link(const name_t&);
+
+    map_t _underlying;
```
can we name `_underlying` to reflect what it's storing? is it `_link_metadata`?

dotnwat commented on src/v/cluster/panda_link/table.h:
```diff
@@ -0,0 +1,70 @@
+private:
+    static constexpr auto accepted_commands = cluster::make_commands_list<
+      cluster::panda_link_upsert_cmd,
+      cluster::panda_link_remove_cmd>();
```
if this moved to the .cc file could we eliminate the commands.h include?

dotnwat commented on src/v/cluster/panda_link/table.cc:
```diff
@@ -0,0 +1,158 @@
+                return table.remove_link(remove.key);
+            });
+      })
+      .then([](std::vector<cluster::errc> results) {
```
this can be plucked off as a separate `co_await` rather than chained

dotnwat commented on src/v/features/feature_table.h:
```diff
@@ -451,6 +452,12 @@ inline constexpr std::array feature_schema{
     feature_spec::available_policy::always,
     feature_spec::prepare_policy::always},
+  feature_spec{
+    release_version::v25_2_1,
```
i'm unfamiliar with what we are delivering in 25.2, but assuming it isn't fully functional, i wonder if we want to introduce some kind of always-in-the-future version in addition to explict_only activation? could some oddities pop up with 25.2 version marking?

michael-redpanda replied:
So something like having

```c++
enum class release_version : int64_t {
    MIN = 3,
    v22_1_1 = MIN,
    v22_1_5 = 4,
    v22_2_1 = 5,
    v22_2_6 = 6,
    v22_3_1 = 7,
    v22_3_6 = 8,
    v23_1_1 = 9,
    v23_2_1 = 10,
    v23_3_1 = 11,
    v24_1_1 = 12,
    v24_2_1 = 13,
    v24_3_1 = 14,
    v25_1_1 = 15,
    v25_2_1 = 16,
    FUTURE = 17
    MAX = FUTURE, // affects the latest_version
};
```

and when we have a new release we update it to

```c++
v25_3_1 = 17,
FUTURE = 18
```
?

michael-redpanda replied:
Alternatively I could also create a `v25_3_1` entry right now

dotnwat replied:
i don't know. i'm just trying to establish first if there is a concern. maybe there isn't it just seemed a bit odd that presumably we are going to use feature gates associated with 25.2.1 perhaps for testing, but then later at least the same set of feature gates will be updated for 25.3.1 but this time for real?

dotnwat reviewed (commented):
quick drive by comments i realize this is going to be refactored per comments on slack.

michael-redpanda commented:
Force push:
* Rebased off dev to get new compiller and clang-format (no functional change)

michael-redpanda commented:
Force push:
* Addresses PR comments
* Moved model info to `src/v/panda_link/model`
* Adjusted namespace names

michael-redpanda commented:
Force push:

* Address PR comments

michael-redpanda commented:
Force push:

* Address merge conflicts

michael-redpanda commented:
Force push:

* Fixed error in rebase

mmaslankaprv commented on src/v/panda_link/model/types.h:
```diff
@@ -0,0 +1,131 @@
+    /// Authentication configuration for the connection
+    std::optional<authn_variant> authn_config;
+    /// Path to certificate file to use
+    ss::sstring cert_file_path;
+    /// Path to key file (when mTLS is in use)
+    ss::sstring key_file_path;
+    /// Path to the CA file to use
+    ss::sstring ca_file_path;
```
can we add a todo here to replace that/or add a buffer representation ?

michael-redpanda replied:
I have filed a PR to add support - is that sufficient?

michael-redpanda replied:
[CORE-11926](https://redpandadata.atlassian.net/browse/CORE-11926)

dotnwat commented on src/v/features/feature_table.h:
```diff
@@ -38,6 +38,7 @@ struct feature_table_snapshot;
 /// only used at runtime.  Therefore it is safe to re-use an integer that
 /// has been made available by another feature being retired.
 enum class feature : std::uint64_t {
+    panda_linking_dr = 1ULL << 1U,
```
we are recycling old features? is this the first time we are doing that?

if so, i think there should be more pomp and circumstance around this... what are the rules, footguns, implications, etc...
